packages/astro/src/core/build/buildPipeline.ts

```diff
@@ -0,0 +1,176 @@
+	 *
+	 * ## Errors
+	 *
+	 * It will throw errors if the previous files can't be found in the file system.
+	 *
+	 * @param staticBuildOptions
+	 */
+	static async retrieveManifest(staticBuildOptions: StaticBuildOptions): Promise<SSRManifest> {
+		const manifestEntryUrl = new URL(
+			'manifest.mjs',
+			staticBuildOptions.settings.config.build.server
+		);
+		const manifest: SSRManifest | undefined = await import(manifestEntryUrl.toString());
+		if (!manifest) {
+			throw new Error(
+				"Astro couldn't find the emitted manifest. This is an internal error, please file an issue."
+			);
+		}
+
+		const renderersEntryUrl = new URL('renderers.mjs', staticBuildOptions.settings.config.outDir);
+		const renderers: SSRLoadedRenderer[] | undefined = await import(renderersEntryUrl.toString());
+		if (!renderers) {
+			throw new Error(
+				"Astro couldn't find the emitted renderers. This is an internal error, please file an issue."
+			);
+		}
+		manifest.renderers = renderers;
+		return manifest;
+	}
+
+	/**
+	 * It collects the routes to generate during the build.
+	 *
+	 * It returns a map of page information and their relative entry point as a string.
+	 */
+	retrieveRoutesToGenerate(): Map<PageBuildData, string> {
+		const pages = new Map<PageBuildData, string>();
+
+		for (const [entryPoint, filePath] of this.#internals.entrySpecifierToBundleMap) {
+			// virtual pages can be emitted with different prefixes:
+			// - the classic way are pages emitted with prefix ASTRO_PAGE_RESOLVED_MODULE_ID -> plugin-pages
+			// - pages emitted using `build.split`, in this case pages are emitted with prefix RESOLVED_SPLIT_MODULE_ID
+			if (
+				entryPoint.includes(ASTRO_PAGE_RESOLVED_MODULE_ID) ||
+				entryPoint.includes(RESOLVED_SPLIT_MODULE_ID)
+			) {
+				const [, pageName] = entryPoint.split(':');
+				const pageData = this.#internals.pagesByComponent.get(
+					`${pageName.replace(ASTRO_PAGE_EXTENSION_POST_PATTERN, '.')}`
+				);
+				if (!pageData) {
+					throw new Error(
+						"Build failed. Astro couldn't find the emitted page from " + pageName + ' pattern'
+					);
+				}
+
+				pages.set(pageData, filePath);
+			}
+		}
+		for (const [path, pageData] of this.#internals.pagesByComponent.entries()) {
+			if (pageData.route.type === 'redirect') {
+				pages.set(pageData, path);
+			}
+		}
+		return pages;
+	}
+
+	async #handleEndpointResult(_: Request, response: EndpointCallResult): Promise<Response> {
+		if (response.type === 'response') {
+			if (!response.response.body) {
+				return new Response(null);
+			}
+			const ab = await response.response.arrayBuffer();
+			const body = new Uint8Array(ab);
+			return new Response(body, {
+				headers: { ...response.response.headers },
+			});
+		} else {
+			if (response.encoding) {
+				const headers = new Headers();
+				headers.set('X-Astro-Encoding', response.encoding);
+				return new Response(response.body, {
+					headers,
+				});
```

Comment:
I thought we could store the `encoding` coming from a `simple` response using an internal header. What do you think? When generating a page, we read it and use it when we write the file to file system.